server/android/src/main/java/dev/slimevr/android/tracking/trackers/hid/TrackersHID.kt

```diff
@@ -0,0 +1,283 @@
+				)
+				usbManager.requestPermission(usbDevice, permissionIntent)
+			}
+		}
+	}
+
+	private fun removeDevice(hidDevice: UsbDevice) {
+		this.devicesByHID[hidDevice]?.let {
+			synchronized(this.devices) {
+				for (id in it) {
+					val device = this.devices[id]
+					for (value in device.trackers.values) {
+						if (value.status == TrackerStatus.OK) {
+							value.status =
+								TrackerStatus.DISCONNECTED
+						}
+					}
+				}
+			}
+
+			this.devicesByHID.remove(hidDevice)
+
+			val oldConn = this.connByHID.remove(hidDevice)
+			val serial = oldConn?.serialNumber ?: "Unknown"
+			oldConn?.close()
+
+			LogManager.info("[TrackerServer] Linked HID device removed: $serial")
+		}
+	}
+
+	private fun deviceIdLookup(hidDevice: UsbDevice, deviceId: Int, deviceName: String? = null, deviceList: MutableList<Int>): HIDDevice? {
+		synchronized(this.devices) {
+			deviceList.map { this.devices[it] }.find { it.hidId == deviceId }?.let { return it }
+			if (deviceName == null) { // not registered yet
+				return null
+			}
+			val device = HIDDevice(deviceId)
+			// server wants tracker to be unique, so use combination of hid serial and full id // TODO: use the tracker "address" instead
+			// TODO: the server should not setup any device, only when the receiver associates the id with the tracker "address" and sends this packet (0xff?) which it will do occasionally
+			// device.name = hidDevice.serialNumber ?: "Unknown HID Device"
+			// device.name += "-$deviceId"
+			device.name = deviceName
+			device.manufacturer = "HID Device" // TODO:
+			// device.manufacturer = hidDevice.manufacturer ?: "HID Device"
+// 			device.hardwareIdentifier = hidDevice.serialNumber // hardwareIdentifier is not used to identify the tracker, so also display the receiver serial
+// 			device.hardwareIdentifier += "-$deviceId/$deviceName" // receiver serial + assigned id in receiver + device address
+			device.hardwareIdentifier = deviceName // the rest of identifier wont fit in gui
+			this.devices.add(device)
+			deviceList.add(this.devices.size - 1)
+			VRServer.instance.deviceManager.addDevice(device) // actually add device to the server
+			LogManager
+				.info(
+					"[TrackerServer] Added device $deviceName for ${hidDevice.serialNumber}, id $deviceId",
+				)
+			return device
+		}
+	}
+
+	private fun dataRead() {
+		synchronized(devicesByHID) {
+			var devicesPresent = false
+			var devicesDataReceived = false
+			val q = intArrayOf(0, 0, 0, 0)
+			val a = intArrayOf(0, 0, 0)
+			val m = intArrayOf(0, 0, 0)
+			for ((hidDevice, deviceList) in devicesByHID) {
+				val dataReceived = ByteArray(64)
+				val conn = connByHID[hidDevice]!!
+				val dataRead = conn.deviceConnection.bulkTransfer(conn.endpointIn, dataReceived, dataReceived.size, 0)
+
+				// LogManager.info("[TrackerServer] HID data read ($dataRead bytes): ${dataReceived.contentToString()}")
+
+				devicesPresent = true // Even if the device has no data
+				if (dataRead > 0) {
+					// Process data
```

Comment:
Can this also be moved to common code?

Reply:
It was kinda a pain because it uses the OS specific USB device class extensively in this part, so it was left OS specific